### ISSUE TITLE
doc.getPageImageList Doesn't return any images, even though they exist

### ISSUE THREAD

**Issue by lightninglu10** [pymupdf/PyMuPDF#464]
[6.pdf](https://github.com/pymupdf/PyMuPDF/files/4326999/6.pdf)

In this PDF Above, images exist in it but doc.getPageImageList always returns an empty list. Is this due to the PDF being malformatted or is something else going on?

**Comment by lightninglu10** [comment]
Okay, maybe this is not a bug, but instead these are not actual images. They look to be some other type of metadata -- anyone know the best way to identify / try to extract them?

**Comment by JorjMcKie** [comment]
> ... or is something else going on?

Something else is going on indeed!
This is a case of something that looks like a duck, walks like a duck, quacks like a duck, but still isn't one 😆.
But seriously: this is a drawing made by using PDF's Postscript-like mini-language. Find the syntax elements in the PDF manual on page 985, "Operator Summary", if you are interested.
All the "images" in this PDF should thus rather be called **_drawings_**: they consist of elementary paint commands like "draw a rectangle here", or "draw a line from p1 to p2", or also "insert this text text here", etc.
There is no way to get a hold of these things if you do not want to basically recreate some type of PDF viewer software yourself. You would have to interpret stuff like this (creates the drawing on page 7, "Figure 1: Words and articles added to Wikipedia since its inception"). I have added a few comments (starting with "%"):

```postscript
1 j  % Set line join style
1 g 0 j 0 w [ ] 0 d 1 G 1 g  % set colors, line width, etc.
0 0 m  % go to point (0, 0)
527.3125 0 l  % from there draw a line to (527.3125, 0)
527.3125 356.340625 l  % from there draw a line to (527.3125, 356.340625)
0 356.340625 l  % from there draw a line to (0, 356.340625)
h  % Close subpath 

f  % Fill path using nonzero winding number rule (which results in something looking like a rectangle)
/A1 gs 0 G

68.35625 56.1375 m
458.95625 56.1375 l
458.95625 327.9375 l
68.35625 327.9375 l
h

f
/A2 gs 0.15 g 1 j 1 w [ ] 0 d 0.15 G 0.15 g
68.35625 56.1375 m
68.35625 50.1375 l

B
/A2 gs
BT /F1 13 Tf
51.825 33.2625 Td
(2001) Tj ET
/A2 gs
118.486731 56.1375 m
118.486731 50.1375 l
...
```

**Comment by JorjMcKie** [comment]
Further information can be extracted however:
This very drawing was imported from some other PDF called ``/compile/graphics/wiki_monthly.pdf`` and has been created using matplotlib and its PDF backend.
Here a are a few PYMuPDF statements which tracked down the information:

```python
>>> print(doc.xrefObject(page.xref))  # print page 7 object definition
<<
  /Contents 14 0 R
  /Group 704 0 R
  /MediaBox [ 0 0 612 792 ]
  /Parent 1004 0 R
  /Resources 701 0 R
  /Type /Page
>>
>>> print(doc.xrefObject(701))  # follow pointers to some other objects:
<<
  /Font <<
    /F113 1001 0 R
    /F95 998 0 R
  >>
  /ProcSet [ /PDF /Text ]
  /XObject <<
    /Im1 13 0 R
  >>
>>
>>> print(doc.xrefObject(13))
<<
  /BBox [ 0 0 527.3125 356.34065 ]
  /Filter /FlateDecode
  /FormType 1
  /Group 704 0 R
  /PTEX.FileName (/compile/graphics/wiki_monthly.pdf)
  /PTEX.InfoDict 705 0 R
  /PTEX.PageNumber 1
  /Resources <<
    /ExtGState <<
      /A1 <<
        /CA 0
        /Type /ExtGState
        /ca 1
      >>
      /A2 <<
        /CA 1
        /Type /ExtGState
        /ca 1
      >>
    >>
    /Font <<
      /F1 706 0 R
    >>
    /Pattern <<
    >>
    /ProcSet [ /PDF /Text /ImageB /ImageC /ImageI ]
    /Shading <<
    >>
    /XObject <<
    >>
  >>
  /Subtype /Form
  /Type /XObject
  /Length 4859
>>
>>> print(doc.xrefObject(704))
<<
  /CS /DeviceRGB
  /S /Transparency
  /Type /Group
>>
>>> print(doc.xrefObject(705))
<<
  /CreationDate (D:20170919100412-04'00')
  /Creator (matplotlib 2.0.2, http://matplotlib.org)
  /Producer (matplotlib pdf backend)
>>
>>> 
```

**Comment by lightninglu10** [comment]
Gotcha! Curious then too -- I see that you've found something here: `doc.xrefObject(13)` on page 7, and I'm wondering how you did so. Are you looping through all of the xrefObjects to look for something specific?

What I was thinking we could do in this case was if we knew where the object was then we can take a "snapshot" of it and save it as an image. 

I guess this is what you're saying about building a PDF reader, but is the above possible? Finding the coordinates of a specific object in the PDF.

**Comment by JorjMcKie** [comment]
Oops, I just realized my last response did not reach you b/o my internet problems.
Ok then, one more time:

1. there is a document method returning XObjects of the page - similar to ``getPageImageList()``. Invoked like this (an internal method currently): ``doc._getPageInfo(pno, 3)``. 3 stands for xobjects.
2. it returns ``(xref, "refname", level, bbox)`` recursively for each xobject referenced. If level == 0, then the page directly references the xobject, otherwise some xobject further down does that. The bbox is currently not correct. It would otherwise tell the location of xobject on the page

Ignited by your inquiry I have made a few changes to PyMuPDF (being published wth the next version), which allow extracting embedded PDF pages like in your example.
My script creates a new PDF where each page consists of an xobject on some input page. The output page size equals the input drawing dimension.
So for your example I am getting this:
[xobj-out.zip](https://github.com/pymupdf/PyMuPDF/files/4331905/xobj-out.zip)

**Comment by lightninglu10** [comment]
Wow that's awesome @JorjMcKie . Do you have an estimate on when that next version would be ready? Would love to use this functionality.

**Comment by JorjMcKie** [comment]
Give me a couple of days, maybe three to four or so.
I just published v1.16.12 and realized that a few corner cases need to be improved / corrected there.
What is your environment? Maybe I can send you a preliminary wheel using this channel ...

**Comment by JorjMcKie** [comment]
I uploaded v1.16.13 which addresses your issue. This is the [script](https://github.com/pymupdf/PyMuPDF-Utilities/blob/master/examples/xobj-extract.py) which produces the example I sent you.
